app/Http/Controllers/AlbumController.php

```diff
@@ -251,8 +251,21 @@ public function getPublic(Request $request)
 					if ($this->sessionFunctions->has_visible_album($album->id)) {
 						return 'true';
 					}
-					if ($album->password == '' || Hash::check($request['password'], $album->password)) {
-						$this->sessionFunctions->add_visible_album($album->id);
+					if (Hash::check($request['password'], $album->password)) {
+						// We add all the albums that the password unlocks so
+						// that the user is not repeatedly asked to enter the
+						// password as they browse through the hierarchy.  This
+						// should be safe as the list of such albums is not
+						// exposed to the user and is considered as the last
+						// access check criteria.
+						$albums = Album::get();
```

Comment:
This should be refined with a `where('password','!=','')`

Reply:
Good point, I don't know what I was thinking when I wrote this...